### ISSUE TITLE
1.1.1-pre1: s_client: PSK ciphersuites no longer usable even with TLS 1.2

### ISSUE THREAD

**Issue by mpg** [openssl/openssl#5406]
Hi,

Usage of PSK ciphersuites with `openssl s_client` appears to be broken with OpenSSL 1.1.1-pre1, or perhaps the way `s_client` should be invoked was changed, but I couldn't find anything about it by looking for PSK in the CHANGES file or the command's help message.

**Steps to reproduce**
1. Start a server with `openssl s_server -cipher NULL,ALL -psk abc123 -nocert` (any version, tested with OpenSSL 1.0.2g and 1.1.1-pre1)
2. Using OpenSSL 1.0.2g, start the client with: `openssl s_client -psk abc123 -cipher PSK-AES128-CBC-SHA` then shut it down. Possible variant: replace the ciphersuite with any other `PSK` ciphersuite (tested with `PSK-ARIA128-GCM-SHA256` too for example).
3a. Using OpenSSL 1.1.1-pre1, start a client with the same command line: `/path/to/openssl-1.1.1-pre1/openssl s_client -psk abc123 -cipher PSK-AES128-CBC-SHA`
3b. Using OpenSSL 1.1.1-pre1, same as step 3a but with `-tls1_2` flag added to the client's command line.

**Observed behaviour**
At step 2, a connection is successfully established.
At step 3a, the client prints the following message on stderr:
```
140108797024000:error:141A90B5:SSL routines:ssl_cipher_list_to_bytes:no ciphers available:ssl/statem/statem_clnt.c:3720:No ciphers enabled for max supported SSL/TLS version
```
At step 3b, the client prints the following message on stderr:
```
140653111523072:error:142120DB:SSL routines:tls_construct_ctos_early_data:bad psk:ssl/statem/extensions_clnt.c:761:
```
At both steps 3a and 3b, the server displays an error message, and observing the connection with wireshark shows that the client sent a fatal alert message instead of a ClientHello.

**Expected behaviour**
At steps 3a and 3b, a connection should be successfully established as in step 2. Alternatively, if it's necessary to modify the way the client is invoked the correct invocation should be clearly documented, preferably with a pointer in the CHANGES file and ideally `s_client` giving a useful hint about it when invoked incorrectly.

Thanks,
Manuel.

**Comment by mattcaswell** [comment]
> `140108797024000:error:141A90B5:SSL routines:ssl_cipher_list_to_bytes:no ciphers available:ssl/statem/statem_clnt.c:3720:No ciphers enabled for max supported SSL/TLS version`

This is a known issue not specific to PSK usage. Basically you are attempting to create a connection with TLSv1.3 enabled, but with no ciphersuites which support TLSv1.3 present. Currently being tracked in #5359. A possible solution to this issue is being investigated in #5392. See also (in particular the ciphersuites section):

https://www.openssl.org/blog/blog/2018/02/08/tlsv1.3/

> `140653111523072:error:142120DB:SSL routines:tls_construct_ctos_early_data:bad psk:ssl/statem/extensions_clnt.c:761:`

This is also a known issue which has already been fixed in #5205.

Also relevant is #5378 which discusses some other compatibility issues with PSK between 1.1.0 and 1.1.1.

As all of the problems raised here are either already fixed or being tracked in other issues, I'm closing this.

**Comment by mpg** [comment]
Thank you very much for your prompt and helpful reply! I built today's master branch and it indeed fixes the TLS 1.2 for me too. So, thanks for the prompt fix too!
